### PR TITLE
Modularize Lexer.x

### LINKED ISSUE #563: Refactor how we use Happy
We use [Alex](https://haskell-alex.readthedocs.io/en/latest/) for generating a lexer for Wasp. Alex allows us to define the lexical grammar (and appropriate actions) in a _Lexer.x_ file using Alex's DSL. This file then compiles to Haskell and can be used as a regular module.

Alex's DSL syntax supports writing and using plain Haskell (e.g., functions, imported modules) inside the same `Lexer.x` file (read [the intro to Alex](https://haskell-alex.readthedocs.io/en/latest/introduction.html) for details).
Unfortunately, IDEs have limited support for Haskell code inside Alex's files. In other words, the best approach is to define all possible helper functions in a regular `.hs` file outside of the _Lexer.x_ file (something like `LexerUtils.hs`) and then import and use them with Alex.

**All lexer utility functions are currently defined inside the _Lexer.x_ file. This implies poor IDE and IntelliSense support. Therefore, we should move them to a regular _LexerUtils.hs_ file and import them inside _Lexer.x_.**

Relevant files:
- _Lexer.x_: https://github.com/wasp-lang/wasp/blob/main/waspc/src/Wasp/Analyzer/Parser/Lexer.x
- _LexerUtils.hs_ - does not yet exist, you should put it here: https://github.com/wasp-lang/wasp/blob/main/waspc/src/Wasp/Analyzer/Parser/

### PR DESCRIPTION
**Pull request by modotte**
# Description

Extract utility functions from Lexer.x into a separate file (LexerUtils.hs).

Fixes #563 

## Type of change

Please select the option(s) that is more relevant.

- [x] Code cleanup
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### CODE REVIEW

**Comment by Martinsos** [comment]
Hey @modotte thanks for creating this PR!!
Looks good! My main question would be, how did you decide which functions go to `LexerUtils` file, and which functions remain in `Lexer`? I ask because I wonder if we could have put more of them into `LexerUtils`? I believe the ones you left in Lexer are the ones for which comments say they are required by `Alex`, but I believe we can move even those to the LexerUtils since we are importing them all in Lexer. So we might also want to move them, in order to get nice IDE features in our .hs file, and we can test it and so on.

I also wonder if we should test these functions now that we have them in LexerUtils.hs -> but we can make a separate issue/PR for that, so we don't bloat this one.

@sodic since you were the author of the original issue let us also know what you think!

**Comment by modotte** [comment]
@Martinsos Hello!

You are definitely correct about more functions that can be moved out of Lexer.x into the new file and I totally missed them based on the "required by Alex" comments. I apologize for that. I'll push a new commit to reflect new changes based on the given review. 

Anyways, I brought in three more functions into LexerUtils.hs in the new commit except for `lexer` itself because of the dependency of `alexScan` which I'm not really sure how to decouple from it yet.

Regarding testing, I think it is ideal to have them implemented ASAP indeed, but perhaps that would require me some more time to try out things before this PR could be actually ready. (I'm quite new to Haskell still :sweat_smile:).

**Comment by Martinsos** [comment]
Cool @modotte ! All makes sense to me. Regarding `lexer` -> we could extract even that one, by making it take `alexScan` as an argument -> maybe we should do that?
I will let @sodic take it from here since he was the latest one touching this part of code!

**Comment by sodic** [comment]
> Cool @modotte ! All makes sense to me. Regarding `lexer` -> we could extract even that one, by making it take `alexScan` as an argument -> maybe we should do that? I will let @sodic take it from here since he was the latest one touching this part of code!

I discussed this with @Martinsos in person. We agreed it's probably best not to do any further extracting. The method is full of Alex-specific stuff (`alexScan`, `AlexError`, `AlexSkip`, etc.). Moving it out of the `Lexer.x` files would make it too complicated.

**Review by sodic — APPROVED** [review]
Thanks for fixing this! I left a single comment, but nothing major.

File: waspc/src/Wasp/Analyzer/Parser/LexerUtils.hs [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+import Control.Monad.State.Lazy (gets)
+import Data.Word (Word8)
+import Codec.Binary.UTF8.String (encodeChar)
+
+type AlexInput = ParserInput
+
+-- Convert the ParserState's start code to an int for Alex to use
+startCodeToInt :: Int -> LexerStartCode -> Int
+startCodeToInt _ DefaultStartCode = 0
+startCodeToInt quoter (QuoterStartCode _) = quoter
```
[the remark this answers is not in the document]
Comment (modotte) [reply waspc/src/Wasp/Analyzer/Parser/LexerUtils.hs:23]: If that's the case, then I probably would say to let this part stay in LexerUtils.hs. While it's small, it also now possible to automate a test for it.

**Review by sodic — COMMENTED** [review]
Hey @modotte, please fix the CI issues and I'll merge the PR :)

You can see the failed CI tests here: https://github.com/wasp-lang/wasp/runs/6574572478?check_suite_focus=true
You can find the details on running code checks locally here: https://github.com/wasp-lang/wasp/tree/main/waspc#code-analysis

**Pull request merged**